make/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+//
+// Copyright (c) 2024 Hemi Labs, Inc.
+//
+// This file is part of the posixutils-rs project covered under
+// the MIT License.  For the full license text, please see the LICENSE
+// file in the root directory of this project.
+// SPDX-License-Identifier: MIT
+//
+
+mod config;
+pub use config::Config;
+mod error_code;
+pub use error_code::ErrorCode;
+
+use std::{
+    env, fs,
+    process::{self, Command},
+    time::SystemTime,
+};
+
+use makefile_lossless::{Makefile, Rule, VariableDefinition};
+use ErrorCode::*;
+
+/// Represents the make utility with its data and configuration. 
+///
+/// The only way to create a `Make` is from a `Makefile` and a `Config`.
+pub struct Make {
+    variables: Vec<VariableDefinition>,
+    rules: Vec<Rule>,
+
+    pub config: Config,
+}
+
+impl Make {
+    /// Retrieves the rule that has the given target.
+    ///
+    /// # Returns
+    ///
+    /// - `Some(rule)` if a rule with the target exists.
+    /// - `None` if no rule with the target exists.
+    pub fn target_rule(&self, target: impl AsRef<str>) -> Option<&Rule> {
+        self.rules
+            .iter()
+            .find(|rule| rule.targets().next().unwrap() == target.as_ref())
```

Comment:
Is unwrapping safe here? Maybe a flat_map could be useful here?

Reply:
It is kinda safe because it's guaranteed on the `makefile-lossless` side, but I will rewrite this just in case.